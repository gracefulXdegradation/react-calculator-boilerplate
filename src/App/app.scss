.DCMN-logo {

    img {
        width: 300px;
    }
}

.button {
    display: inline-flex;
    width: 20px;
    height: 20px;
    justify-content: center;
    align-items: center;
}
