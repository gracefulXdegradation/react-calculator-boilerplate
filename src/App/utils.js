export const getDecimalPoints = num => (`${num}`.split('.')[1] || '').length;

export const getMaxDecimalPoints = (numA, numB) => Math.max(getDecimalPoints(numA), getDecimalPoints(numB));

export const formatNumber = (num) => {
    const hasDot = num.includes('.');
    const [int, dec] = num.split('.');
    const formattedInt = int.split('').reverse().reduce(
        (acc, val, i) => acc + (i > 0 && i % 3 === 0 ? ' ' : '') + val, '',
    ).split('')
        .reverse()
        .join('');

    return formattedInt + (hasDot ? `.${dec}` : '');
};
