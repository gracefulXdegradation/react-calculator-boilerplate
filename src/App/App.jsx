import React, { useState } from 'react';
import './app.scss';

const operations = {
    '+': (a, b) => a + b,
    '-': (a, b) => a - b,
};

const getDecimalPoints = num => (`${num}`.split('.')[1] || '').length;
const getMaxDecimalPoints = (numA, numB) => Math.max(getDecimalPoints(numA), getDecimalPoints(numB));
const execute = (numA, numB, op) => {
    const result = operations[op](parseFloat(numA), parseFloat(numB));

    return result.toFixed(getMaxDecimalPoints(numA, numB));
};
const formatNumber = (num) => {
    const hasDot = num.includes('.');
    const [int, dec] = num.split('.');
    const formattedInt = int.split('').reverse().reduce(
        (acc, val, i) => acc + (i > 0 && i % 3 === 0 ? ' ' : '') + val, '',
    ).split('')
        .reverse()
        .join('');

    return formattedInt + (hasDot ? `.${dec}` : '');
};

const maxInputReached = num => num.replace('.', '').length >= 9;

const App = () => {
    const [screen, setScreen] = useState('0');
    const [memo, setMemo] = useState(null);
    const [op, setOp] = useState(null);

    const pushNumButton = (value) => {
        if (op && memo === null) {
            setMemo(screen);
            setScreen(`${value}`);
        } else {
            if (maxInputReached(screen)) {
                return;
            }
            const newScreen = `${screen}${value}` * 1; // get rid of leading spaces

            setScreen(`${newScreen}`);
        }
    };

    const pushDotButton = () => {
        if (maxInputReached(screen)) {
            return;
        }
        if (op && memo === null) {
            setMemo(screen);
            setScreen('0.');
        }
        if (!screen.includes('.')) {
            setScreen(`${screen}.`);
        }
    };

    const pushEquals = () => {
        if (op) {
            const result = execute(memo || 0, screen, op);

            setScreen(`${result}`);
            setMemo(null);
            setOp(null);
        }
    };

    const setOperation = (operation) => {
        if (op) {
            pushEquals();
        }
        setOp(operation);
    };

    const clear = () => {
        setScreen('0');
        setMemo(null);
        setOp(null);
    };

    return (
        <div>
            <div>{formatNumber(screen)}</div>
            {[0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(value => (
                <button
                    type="button"
                    className="button"
                    key={`button-${value}`}
                    onClick={() => pushNumButton(value)}
                >
                    {value}
                </button>
            ))}
            <button type="button" className="button" onClick={pushDotButton}>.</button>
            <button type="button" className="button" onClick={() => setOperation('+')}>+</button>
            <button type="button" className="button" onClick={() => setOperation('-')}>-</button>
            <button type="button" className="button" onClick={pushEquals}>=</button>
            <button type="button" className="button" onClick={clear}>AC</button>
        </div>
    );
};

export default App;
