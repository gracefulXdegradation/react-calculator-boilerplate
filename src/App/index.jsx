import React from 'react';
import Calc from '../components/calc';
import './app.scss';

const App = () => (
    <div className="app">
        {/* <div className="DCMN-logo">
            <img src="https://dcmn.com/assets/dcmn-logo-redesign-black.svg" alt="DCMN logo" />
        </div> */}
        <Calc />
    </div>
);

export default App;
