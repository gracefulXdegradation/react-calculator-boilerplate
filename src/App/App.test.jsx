import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import App from './App';

const getScreen = wrapper => wrapper.find('[data-role="screen"]');
const clickButton = (button, wrapper) => wrapper.find(`[data-role="button_${button}"]`).simulate('click');
const pressKey = key => act(() => {
    document.body.dispatchEvent(new KeyboardEvent('keypress', { key }));
});
const pressKeys = keys => keys.split('').forEach(key => pressKey(key));
const clickButtons = (buttons, wrapper) => buttons.split('').forEach(button => clickButton(button, wrapper));

describe('App', () => {
    let consoleSpy;
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<App />);
    });

    beforeAll(() => {
        consoleSpy = jest.spyOn(console, 'error');
    });

    it('renders the App component and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('renders logo', () => {
        expect(wrapper.find('img')).toHaveLength(1);
    });

    it('shows no errors on console', () => {
        expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('shows 0 as init value', () => {
        const screen = getScreen(wrapper);

        expect(screen).toHaveLength(1);
        expect(screen.props().value).toEqual('0');
    });

    it('handles number clicks correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('123', wrapper);
        expect(screen.getDOMNode().value).toEqual('123');
    });

    it('handles decimal numbers input correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('0.002300', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.002300');
    });

    it('allows max one decimal separator', () => {
        const screen = getScreen(wrapper);

        clickButton('.', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.');
        clickButtons('123.4', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.1234');
    });

    it('groups digits in integer part correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('1234', wrapper);
        expect(screen.getDOMNode().value).toEqual('1 234');
        clickButton('5', wrapper);
        expect(screen.getDOMNode().value).toEqual('12 345');
        clickButtons('.6789', wrapper);
        expect(screen.getDOMNode().value).toEqual('12 345.6789');
    });

    it('trims trailing zeros correctly: int', () => {
        const screen = getScreen(wrapper);

        clickButtons('.1+1.9=', wrapper);
        expect(screen.getDOMNode().value).toEqual('2');
    });

    it('trims trailing zeros correctly: float', () => {
        const screen = getScreen(wrapper);

        clickButtons('.002+1.898=', wrapper);
        expect(screen.getDOMNode().value).toEqual('1.9');
    });

    it('handles key press correctly', () => {
        const screen = getScreen(wrapper);

        pressKeys('1+1.9=');
        expect(screen.getDOMNode().value).toEqual('2.9');
    });
});
