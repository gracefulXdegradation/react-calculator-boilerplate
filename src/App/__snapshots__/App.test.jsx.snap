// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the App component and matches snapshot 1`] = `
<form>
  <div
    className="DCMN-logo"
  >
    <img
      alt="DCMN logo"
      src="https://dcmn.com/assets/dcmn-logo-redesign-black.svg"
    />
  </div>
  <input
    readOnly={true}
    type="text"
    value="0"
  />
  <button
    className="button"
    key="button-0"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className="button"
    key="button-1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="button"
    key="button-2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="button"
    key="button-3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="button"
    key="button-4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="button"
    key="button-5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="button"
    key="button-6"
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="button"
    key="button-7"
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className="button"
    key="button-8"
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className="button"
    key="button-9"
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    =
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
</form>
`;
